common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) map { actualEntry =>
+                // ... and see what we get back
+
+                if (transid == actualEntry.transid) {
+                    //
+                    // then we won the race, and now own the entry
+                    //
+                    logger.debug(this, s"Write initiated $key")
+                    listenForWriteDone(key, actualEntry, future)
+
+                } else {
+                    //
+                    // then either we lost the race, or there is an operation in progress on this key
+                    //
+                    // note that we don't care what that other operation is doing; unlike for lookups,
+                    // we don't need to do anything special (e.g. coalescing, as in the case of lookups)
+                    //
+                    logger.debug(this, s"Write under ${actualEntry.state.get} $key")
+                }
+            }
+
+        } else {
+            // not caching
+        }
+
+        future
+    }
+
+    def cacheSize: Int = cache.size
+
+    /**
+     * Log a cache hit
+     *
+     */
+    private def makeNoteOfCacheHit(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_HIT, s"[GET] serving from cache: $key")(logger)
+    }
+
+    /**
+     * Log a cache miss
+     *
+     */
+    private def makeNoteOfCacheMiss(key: Any)(implicit transid: TransactionId, logger: Logging) {
+        transid.mark(this, LoggingMarkers.DATABASE_CACHE_MISS, s"[GET] serving from datastore: $key")(logger)
+    }
+
+    /**
+     * We have initiated a read (in cacheLookup), now handle its completion
+     *
+     */
+    private def listenForReadDone(key: Any, entry: Entry, future: => Future[W], promise: Promise[W])(
+        implicit logger: Logging, ec: ExecutionContext): Unit = {
+
+        future onComplete {
+            case Success(value) => {
+                //
+                // if the datastore read was successful, then transition to the Cached state
+                //
+                logger.debug(this, "Read backend part done, now marking cache entry as done")
+                if (entry.readDone()) {
+                    promise success value
+                } else {
+                    // if this ever happens, this cache impl is buggy
+                    readOops(key, entry, promise, ConcurrentOperationUnderRead(entry.state.get))
+                }
+            }
+            case Failure(t) => {
+                //
+                // oops, the datastore read failed. invalidate the cache entry
+                //
+                // note that this might be a perfectly legitimate failure,
+                // e.g. a lookup for a non-existant key; we need to pass the particular t through
+                //
+                readOops(key, entry, promise, t)
+            }
+        }
```

Comment:
The logic here can be simplified:

```
future.flatMap { value =>
    if (entry.readDone()) {
        Future.successful(value)
    } else {
        Future.failed(ConcurrentOperationUnderRead(entry.state.get))
    }
} onFailure {
    case t =>
        entry.invalidate
        cache remove key
        promise failure t
}
```

With the simplification in place, note that i inlined `readOops` because it'ld be only called once anyway. Also we could then inline `listenForReadDone` above, putting the relevant logic in one place, preventing us from passing a `Promise` through different methods. What do you think?